backend/src/controllers/userController.ts

```diff
@@ -30,42 +60,190 @@ export const createUser = asyncHandler(async (req, res, next) => {
+    }
+  }
+
+  res.status(200).json(responseData);
 });
 
 // @desc Update a user
 // @route PATCH /api/users/:id
 // @access Private
 export const updateUser = asyncHandler(async (req, res, next) => {
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    return next(createHttpError(400, validationErrorParser(errors)));
+  }
+
   const { id } = req.params;
-  const {
-    type,
-    linkedIn,
-    phoneNumber,
-    major,
-    classLevel,
-    company,
-    shareProfile,
-  } = req.body;
+  const validatedData = matchedData(req, { locations: ["body"] });
+
+  // check if at least one field to update is in body
+  if (Object.keys(validatedData).length === 0) {
+    return next(createHttpError(400, "At least one field required to update."));
+  }
+
+  // find the user to update
+  const foundUser = await User.findByIdAndUpdate(
+    id,
+    { $set: validatedData },
+    { new: true, runValidators: true },
+  );
+
+  // check if the user exists
+  if (!foundUser) {
+    return next(createHttpError(404, "User not found."));
+  }
+
+  res.status(200).json(foundUser);
 });
 
 // @desc Delete a user
 // @route DELETE /api/users/:id
 // @access Private
 export const deleteUser = asyncHandler(async (req, res, next) => {
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    return next(createHttpError(400, validationErrorParser(errors)));
+  }
+
   const { id } = req.params;
+
+  const foundUser = await User.findByIdAndDelete(id);
+
+  // check if the user already exists
+  // check if the user exists
+  if (!foundUser) {
+    return next(createHttpError(404, "User not found."));
+  }
+
+  res.status(200).json(foundUser);
 });
 
 // @desc Get alumni willing to share profile
 // @route GET /api/users/alumni
 // @access Private
 export const getOpenAlumni = asyncHandler(async (req, res, next) => {
-  const { page, perPage, query } = req.query;
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    return next(createHttpError(400, validationErrorParser(errors)));
+  }
+
+  const { page = "1", perPage = "10", query } = req.query;
+  console.log(req.query);
+
+  const pageString = typeof page === "string" ? page : "1";
+  const perPageString = typeof perPage === "string" ? perPage : "10";
+
+  // check if the query parameter is missing or empty
+  if (typeof query !== "string" || query.trim() === "") {
+    return next(createHttpError(400, "Missing or empty query parameter."));
+  }
+
+  const pageNum = parseInt(pageString, 10) || 1;
+  const perPageNum = parseInt(perPageString, 10) || 10;
+
+  const dbQuery = User.find({
+    type: "Alumni",
+    shareProfile: true,
+  });
+
+  // add a name search filter if provided
+  if (query) {
+    dbQuery.where("name").regex(new RegExp(query, "i"));
+  }
+
+  // count total results for pagination
+  const total = await dbQuery.clone().countDocuments();
+
+  // apply pagination
+  dbQuery.skip((pageNum - 1) * perPageNum).limit(perPageNum);
```

Comment:
mb for not specifying but page would already be 0-indexed so no need to subtract it by 1